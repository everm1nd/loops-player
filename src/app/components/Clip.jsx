import React from 'react';
import PropTypes from "prop-types"
import styled from 'styled-components'
import Tone from "tone";
import 'components/Clip/play-icon.svg'
import 'components/Clip/stop-icon.svg'

const assetPath = url => process.env.PUBLIC_URL + url;

const Progress = styled.div`
  animation-duration: ${props => Tone.Time(props.duration).toSeconds()}s !important
`

class Clip extends React.Component {
  constructor(props) {
    super(props)
    this.player = new Tone.Player(assetPath(props.url)).toMaster();
    this.player.loop = true;

    this._togglePlayback = this._togglePlayback.bind(this)
    this.handleClick = this.props.onClick.bind(null, this.props.id)
  }

  _togglePlayback() {
    switch(this.props.playbackState) {
      case 'stopped':
        this.player.stop()
        break;
      case 'started':
        if (this.player.state !== "started") this.player.start('@0.01')
        break;
      default:
        break;
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.props.playbackState !== prevProps.playbackState) this._togglePlayback()
  }

  render() {
    return <button className={`button clip ${this.props.playbackState}`} onClick={this.handleClick}>
      <Progress duration={this.props.duration} className="progress" />
    </button>
  }
}

Clip.defaultProps = {
  playbackState: 'stopped'
}

Clip.propTypes = {
  id: PropTypes.number.isRequired,
  url: PropTypes.string.isRequired,
  playbackState: PropTypes.string,
  onClick: PropTypes.func.isRequired,
  duration: PropTypes.string.isRequired
}

export default Clip;
