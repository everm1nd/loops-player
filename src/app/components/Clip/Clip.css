.clip {
  display: block;
  width: 100px;
  height: 100px;
  background: lightgrey url('play-icon.svg') no-repeat center;
  background-size: 80px;
  cursor: pointer;
  margin-bottom: 2px;
  padding: 0;
}

@keyframes filler {
  from { width: 0%; }
  to { width: 100%; }
}

.clip .progress {
  width: 0;
  height: 100%;
  background: white;
  opacity: 0.3;
}

@keyframes blinker {
  50% { opacity: 0.5; }
}

.clip.starting, .clip.stopping {
  animation: blinker 1.5s ease-in-out infinite;
}

.clip.started .progress {
  animation: filler 4s linear infinite;
}

.clip.stopping .progress {
  animation: filler 4s linear infinite;
}

.clip.starting, .clip.started {
  background: lightgrey url('stop-icon.svg') no-repeat center;
  background-size: 60px !important;
}
