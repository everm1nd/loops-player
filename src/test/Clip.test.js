import React from 'react';
import { shallow } from 'enzyme';
import Clip from 'components/Clip';
import Tone from 'tone';

describe("Clip", () => {
  const wrapper = shallow(<Clip id={0} url="something.wav" onClick={jest.fn()} />);

  it('renders a clip', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("plays when props.playbackState == started", () => {
    wrapper.setProps({ playbackState: "started" })
    expect(wrapper.instance().player.start).toHaveBeenCalled();
  })

  it("stops when props.playbackState == stopped", () => {
    wrapper.setProps({ playbackState: "stopped" })
    expect(wrapper.instance().player.stop).toHaveBeenCalled();
  })
})
