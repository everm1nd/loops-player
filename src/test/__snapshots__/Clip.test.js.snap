// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clip renders a clip 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Clip
    id={0}
    onClick={[Function]}
    playbackState="stopped"
    url="something.wav"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "className": "button clip",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "className": "button clip",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
